### PR TITLE
*: Integrate timeline tracing with TiKV

### PR DESCRIPTION
**Pull request by zhongzc**
Signed-off-by: zhongzc <zhongzc_arch@outlook.com>

### Create a Work Directory

```sh
mkdir trace-tidb && cd trace-tidb
```

### Build TiDB

```sh
git clone -b tracing https://github.com/zhongzc/tidb --depth=1
(cd tidb && make)
```

### Build TiKV

```sh
git clone -b new-minitrace https://github.com/zhongzc/tikv --depth=1
(cd tikv && make build)
```

### Create Config Files

```sh
echo -e '[log]\nslow-threshold=5\n[trace]\nenable = true\njaeger-thrift-compact-agent = "127.0.0.1:6831"\ndatadog-agent = "127.0.0.1:8126"' > tidb.config
```

### Run Jaeger

```sh
docker run --rm -d -p6831:6831/udp -p16686:16686 --name jaeger jaegertracing/all-in-one:latest
```

> You can then navigate to http://localhost:16686 to access the Jaeger UI.

### Set Up Datadog (alternative)

Follow https://docs.datadoghq.com/getting_started/tracing/

### Run TiUP to Start a Cluster Locally

```sh
tiup playground v4.0.12 --db.binpath=tidb/bin/tidb-server --db.config=tidb.config --kv.binpath=tikv/target/debug/tikv-server --tiflash 0
```

### Examples

Trace data of SQL whose cost time more than **`log.slow-threshold`** (set to **5ms** above) will be recorded.

Select:

From Jaeger UI:

Screenshot from 2020-11-26 17-22-21

From Datadog UI:
Screenshot from 2020-12-04 10-59-18

Insert:

From Jaeger UI:
Screenshot from 2020-11-26 17-23-32

From Datadog UI:
Screenshot from 2020-12-04 11-00-08

Screenshot from 2020-12-18 09-42-15

### Release note 
- TBD

### CODE REVIEW

**Comment by baurine** [comment]
Some supplements:

If you install the datadog-agent in a vagrant VM instead of host self according to the link of https://docs.datadoghq.com/getting_started/tracing/ , you need to modify the Vagrantfile, set a private IP for the VM, for example:

```
config.vm.network "private_network", ip: "10.0.1.2"
```

Use `vagrant reload` to restart the VM by the new configuration.

Then change the configuration in tidb.config:

```
datadog-agent = "10.0.1.2:8126"
```

File: trace/reporter.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+}
+
+type converter struct {
+	jTraces   *[]jaeger.Trace
+	dSpanList *datadog.SpanList
+
+	curTraceID          uint64
+	curServiceName      string
+	curJTrace           *jaeger.Trace
+	curSpanIDPrefix     uint32
```
Comment (zhongzc) [inline trace/reporter.go:176]: @breeswish

File: server/conn.go [hunk 2]
```diff
@@ -817,8 +829,12 @@ func (cc *clientConn) Run(ctx context.Context) {
 			err1 := cc.writeError(ctx, err)
 			terror.Log(err1)
 		}
-		cc.addMetrics(data[0], startTime, err)
+		cc.addMetrics(cmd, startTime, err)
 		cc.pkt.sequence = 0
+
+		if traceEnabled {
+			tidbtrace.Report(handle)
```
Comment (zhongzc) [inline server/conn.go:836]: @breeswish

File: server/conn.go [hunk 3]
```diff
@@ -789,9 +792,18 @@ func (cc *clientConn) Run(ctx context.Context) {
 		}
 
 		startTime := time.Now()
+
+		ctx := ctx
+		var handle minitrace.TraceHandle
+		traceEnabled := tidbtrace.Enable.Load()
+		if traceEnabled {
+			ctx, handle = minitrace.StartRootSpan(ctx, cc.packetToSQL(data), rand.Uint64(), &tidbtrace.Context{})
```
Comment (zhongzc) [inline server/conn.go:800]: @c4pt0r You can replace `rand.Uint64()` with another meaningful trace ID here.

Comment (c4pt0r) [reply server/conn.go:800]: Sure

**Comment by tisonkun** [comment]
Closed as `store/tikv` moved to [client-go](https://github.com/tikv/client-go).

Thanks for your contribution @zhongzc ! You may create another PR on both tidb and client-go for reimplementing it.

**Comment by Defined2014** [comment]
Seems it's a very useful PR, I think we should support/merge it.

**Pull request closed without merging**